File: cses/04_graph_algorithms/1686_coin-collector/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần cho nhanh.
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); pos += 1
    m = int(data[pos]); pos += 1

    # Số xu tại mỗi phòng (đánh số 1..n).
    coins = [0] * (n + 1)
    for i in range(1, n + 1):
        coins[i] = int(data[pos]); pos += 1

    # Lưu cạnh và đếm bậc ra để dựng danh sách kề dạng CSR.
    ea = [0] * m          # đỉnh đầu mỗi cạnh
    eb = [0] * m          # đỉnh cuối mỗi cạnh
    off = [0] * (n + 2)   # CSR offsets
    for e in range(m):
        a = int(data[pos]); b = int(data[pos + 1]); pos += 2
        ea[e] = a
        eb[e] = b
        off[a + 1] += 1
    for i in range(1, n + 2):
        off[i] += off[i - 1]

    # Điền các đỉnh kề vào mảng CSR liên tục.
    adj = [0] * m
    fill = off[:]  # con trỏ điền tạm
    for e in range(m):
        a = ea[e]
        adj[fill[a]] = eb[e]
        fill[a] += 1

    # ---- Bước 1: Tarjan SCC (iterative, dùng stack tường minh tránh đệ quy sâu) ----
    index_arr = [-1] * (n + 1)
    low = [0] * (n + 1)
    comp = [-1] * (n + 1)
    on_stack = bytearray(n + 1)
    ptr = off[:]  # con trỏ cạnh hiện tại cho mỗi đỉnh
    sccstack = []
    idx_counter = 0
    ncomp = 0

    for s in range(1, n + 1):
        if index_arr[s] != -1:
            continue
        work = [s]  # stack mô phỏng đệ quy DFS
        index_arr[s] = idx_counter
        low[s] = idx_counter
        idx_counter += 1
        sccstack.append(s)
        on_stack[s] = 1
        while work:
            v = work[-1]
            ei = ptr[v]
            end = off[v + 1]
            recurse = False
            while ei < end:
                w = adj[ei]
                ei += 1
                if index_arr[w] == -1:
                    # Cạnh cây: đi sâu xuống đỉnh chưa thăm.
                    ptr[v] = ei
                    index_arr[w] = idx_counter
                    low[w] = idx_counter
                    idx_counter += 1
                    sccstack.append(w)
                    on_stack[w] = 1
                    work.append(w)
                    recurse = True
                    break
                elif on_stack[w] and index_arr[w] < low[v]:
                    # Cạnh ngược tới đỉnh còn trên stack: cập nhật low.
                    low[v] = index_arr[w]
            if recurse:
                continue
            ptr[v] = ei
            if low[v] == index_arr[v]:
                # v là gốc SCC: pop toàn bộ thành phần và gán id.
                while True:
                    w = sccstack.pop()
                    on_stack[w] = 0
                    comp[w] = ncomp
                    if w == v:
                        break
                ncomp += 1
            work.pop()
            # Truyền low của con lên cha.
            if work:
                pv = work[-1]
                if low[v] < low[pv]:
                    low[pv] = low[v]

    # ---- Bước 2: Giá trị mỗi SCC = tổng xu các phòng bên trong ----
    comp_val = [0] * ncomp
    for v in range(1, n + 1):
        comp_val[comp[v]] += coins[v]

    # ---- Bước 3: Danh sách kề trên DAG condense (chỉ giữ cạnh giữa hai SCC khác nhau) ----
    cond_adj = [[] for _ in range(ncomp)]
    for v in range(1, n + 1):
        cv = comp[v]
        for idx in range(off[v], off[v + 1]):
            cw = comp[adj[idx]]
            if cv != cw:
                cond_adj[cv].append(cw)

    # ---- Bước 4: DP đường đi tổng lớn nhất. Id tăng dần là thứ tự topo đảo ngược,
    # mọi successor có id nhỏ hơn nên đã được tính dp trước. ----
    dp = [0] * ncomp
    ans = 0
    for c in range(ncomp):
        best = 0
        for cw in cond_adj[c]:
            d = dp[cw]
            if d > best:
                best = d
        cur = comp_val[c] + best
        dp[c] = cur
        if cur > ans:
            ans = cur

    sys.stdout.write(str(ans) + "\n")


main()

File: cses/04_graph_algorithms/1686_coin-collector/solution.cpp
#include <algorithm>
#include <iostream>
#include <utility>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, m;
    std::cin >> n >> m;

    // Số xu tại mỗi phòng.
    std::vector<long long> coins(n);
    for (long long& value : coins) {
        std::cin >> value;
    }

    // Danh sách kề adj cho Tarjan, đồng thời lưu lại toàn bộ cạnh gốc để condense.
    std::vector<std::vector<int>> adj(n);
    std::vector<std::pair<int, int>> edges;
    edges.reserve(m);
    for (int i = 0; i < m; ++i) {
        int a, b;
        std::cin >> a >> b;
        --a;
        --b;
        adj[a].push_back(b);
        edges.emplace_back(a, b);
    }

    // Bước 1: Tìm SCC bằng Tarjan iterative (stack tường minh, không đệ quy).
    // Một SCC được đánh số khi nó bị pop, nên mọi cạnh giữa các SCC luôn đi từ
    // id lớn sang id nhỏ (thứ tự topo đảo ngược).
    std::vector<int> index(n, -1);
    std::vector<int> low(n, 0);
    std::vector<int> component(n, -1);
    std::vector<std::size_t> next_edge(n, 0);  // con trỏ cạnh đang xét của mỗi đỉnh
    std::vector<char> on_stack(n, false);
    std::vector<int> scc_stack;  // stack các đỉnh của SCC đang mở
    std::vector<int> dfs_stack;  // stack mô phỏng đệ quy DFS
    scc_stack.reserve(n);
    dfs_stack.reserve(n);

    int timer = 0;
    int component_count = 0;
    for (int start = 0; start < n; ++start) {
        if (index[start] != -1) {
            continue;
        }

        index[start] = low[start] = timer++;
        on_stack[start] = true;
        scc_stack.push_back(start);
        dfs_stack.push_back(start);

        while (!dfs_stack.empty()) {
            const int node = dfs_stack.back();
            if (next_edge[node] < adj[node].size()) {
                const int next = adj[node][next_edge[node]++];
                if (index[next] == -1) {
                    // Cạnh cây: đi sâu xuống đỉnh chưa thăm.
                    index[next] = low[next] = timer++;
                    on_stack[next] = true;
                    scc_stack.push_back(next);
                    dfs_stack.push_back(next);
                } else if (on_stack[next]) {
                    // Cạnh ngược tới đỉnh còn trên stack: cập nhật low.
                    low[node] = std::min(low[node], index[next]);
                }
                continue;
            }

            // Đã duyệt hết cạnh của node: quay lui.
            dfs_stack.pop_back();
            if (low[node] == index[node]) {
                // node là gốc của một SCC: pop toàn bộ thành phần.
                while (true) {
                    const int member = scc_stack.back();
                    scc_stack.pop_back();
                    on_stack[member] = false;
                    component[member] = component_count;
                    if (member == node) {
                        break;
                    }
                }
                ++component_count;
            }
            // Truyền low của con lên cha.
            if (!dfs_stack.empty()) {
                const int parent = dfs_stack.back();
                low[parent] = std::min(low[parent], low[node]);
            }
        }
    }

    // Bước 2: Giá trị mỗi SCC = tổng xu các phòng thuộc SCC đó.
    std::vector<long long> component_coins(component_count, 0);
    for (int room = 0; room < n; ++room) {
        component_coins[component[room]] += coins[room];
    }

    // Bước 3: Danh sách kề của DAG condense (chỉ giữ cạnh nối hai SCC khác nhau).
    std::vector<std::vector<int>> condensed(component_count);
    for (const auto& [from, to] : edges) {
        const int from_component = component[from];
        const int to_component = component[to];
        if (from_component != to_component) {
            condensed[from_component].push_back(to_component);
        }
    }

    // Bước 4: DP đường đi tổng lớn nhất trên DAG. Vì id SCC theo thứ tự topo đảo
    // ngược, mọi successor của c có id nhỏ hơn nên đã được tính trước khi tới c.
    std::vector<long long> best(component_count, 0);
    long long answer = 0;
    for (int c = 0; c < component_count; ++c) {
        long long best_successor = 0;
        for (const int next : condensed[c]) {
            best_successor = std::max(best_successor, best[next]);
        }
        best[c] = component_coins[c] + best_successor;
        answer = std::max(answer, best[c]);
    }

    std::cout << answer << '\n';
    return 0;
}
